Add optional deployment and service for DataDog

ivanilves opened the pull request:
Hi all! :wave: 

## Why do we need it?
* Serve applications that want to send custom metrics with an accessible DogStasD service endpoint.
* Make DaemonSet optional (enabled by default though), so we could put a Datadog deployment into every namespace w/o overhead of re-adding a new DaemonSet (mostly for single-pod deployments). 
* This could also serve, if you have a single Datadog agent pod in every namespace, and you run custom python checks against your services (e.g. HTTP smokes) from this pod and you do not really need a DaemonSet for it, especially if you need a single pod deployment to save resources and avoid duplication of metrics/events (not only K8s ones, think about your custom ones also).   

## GIF
*That's my first chart contribution* :wink:
 
![](https://media.giphy.com/media/CovFciJgWyxUs/giphy.gif)

ivanilves committed:
Modify deployment and added service for DataDog

ivanilves commented:
Signed!

prydonius commented:
Thanks @ivanilves! cc @gtaylor @irabinovitch

apenney commented:
This is fantastic, I was trying to figure out how to do this, you're a hero.

apenney commented:
Just to confirm it works great:

```
Release "datadog" has been upgraded. Happy Helming!
LAST DEPLOYED: Fri May 19 13:41:11 2017
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1/ConfigMap
NAME                      DATA  AGE
datadog-datadog-checksd   0     18d
datadog-datadog-autoconf  0     18d
datadog-datadog-confd     0     18d

==> v1/Service
NAME             CLUSTER-IP      EXTERNAL-IP  PORT(S)   AGE
datadog-datadog  100.68.108.121  <none>       8125/UDP  1m

==> v1beta1/DaemonSet
NAME             DESIRED  CURRENT  READY  UP-TO-DATE  AVAILABLE  NODE-SELECTOR  AGE
datadog-datadog  7        7        1      7           1          <none>         1s

==> v1beta1/Deployment
NAME             DESIRED  CURRENT  UP-TO-DATE  AVAILABLE  AGE
datadog-datadog  1        1        1           0          1s

==> v1/Secret
NAME             TYPE    DATA  AGE
datadog-datadog  Opaque  1     18d


NOTES:
DataDog agents are spinning up on each node in your cluster. After a few
minutes, you should see your agents starting in your event stream:

    https://app.datadoghq.com/event/stream

No further action should be required.```

irabinovitch commented on stable/datadog/templates/service.yaml:
```diff
@@ -0,0 +1,26 @@
+  - port: 8126
+    name: traceport
+    protocol: TCP
+  - port: 7777
```
I would exclude this port 7777 isn't in use and will be dropped in a release coming soon.

irabinovitch commented on stable/datadog/values.yaml:
```diff
@@ -4,6 +4,14 @@ image:
   tag: latest
   pullPolicy: IfNotPresent
 
+# Apart from DaemonSet, also deploy DataDog agent single pod and related service to:
```
NIT: Datadog

irabinovitch commented:
@ivanilves Thanks for working on this. I'm tagging in @hkaj for his thoughts.

A few things to keep in mind:
- We have a release of dd-agent scheduled that automatically handles the event piece. 
- Ideally DogStatsD metrics should be routed to a local DogStatsD  instance rather than cross, otherwise you'll end up with different host tags than you're expecting. Also StatsD and DogStatsD are UDP.

Eventually this would be better served with https://github.com/kubernetes/kubernetes/issues/28610

ivanilves commented:
@Fsero @prydonius @apenney @irabinovitch thanks for your prompt reaction and support :clap: 

@irabinovitch your comments are great!

So I dropped port `7777` out of service mapping and renamed "DataDog" to "Datadog" :smile: 

> We have a release of dd-agent scheduled that automatically handles the event piece.

Awesome! Datadog is being ahead of my expectations :clap: :dancer: 

> Ideally DogStatsD metrics should be routed to a local DogStatsD instance rather than cross, otherwise you'll end up with different host tags than you're expecting. Also StatsD and DogStatsD are UDP.

Yes! We have both setups available, depending on the application.

The local DogStatsD benefits are obvious, but sometimes you need to provide every team (read: every namespace) with an exclusive DogStatsD, not tied to any host, armed with team-specific API key and a set of specific [python] checks run by team. Running a Datadog per pod - resource waste IMO. Running on every node - better, still does not correlate with our "namespace = team account" logic.

For this kind of setups, teams do write metric sending thing by themselves, so they have control on necessary tag "injection" and are aware of possible losses due to the stateless nature of UDP. :wink: 

What we would like to achieve with this PR is to make this awesome Datadog chart even more flexible, while preserving its reasonable default behavior. _I also try to escape vendoring it from our private repo._

Thanks!

**P.S.:**
Great info on the structural changes in https://github.com/kubernetes/kubernetes/issues/28610 though that PR is pretty old and still not merged, which makes me a little less optimistic about when we could enjoy the benefits it will provide. Anyway, thumbed it up!

vreon commented:
FYI this overlaps with kubernetes/charts#722, but the Service from this PR would still be really nice to have

ivanilves commented:
Yes @vreon the overlap does exist, with this PR having a slightly wider scope (for good or for bad, dunno).

I think now @seanknox will merge https://github.com/kubernetes/charts/pull/722 into master and then I could try to rebase against a new master to get Service resource.

seanknox commented:
Hey @ivanilves #722 just landed in master if you want to rebase and/or adjust this PR scope.

ivanilves commented:
Great! @seanknox @vreon I've rebased / readjusted scope of this PR.

Now it's concentrated on deployment option flexibility rather than on K8s events collection.
Thanks @fiunchinho for taking care of the events part though. This made my scope narrower.

fiunchinho commented on stable/datadog/templates/deployment.yaml:
```diff
@@ -1,25 +1,37 @@
         resources:
 {{ toYaml .Values.resources | indent 12 }}
+        ports:
+        - containerPort: 8125
```
This deployment is just to collect events, right? Should this deployment expose the ports and configuration options? What do you think?

ivanilves replied:
No @fiunchinho not only :wink: 

We could use ports with:

> Serve applications that want to send custom metrics with an accessible DogStasD service

and configuration options:

> This could also serve, if you have a single Datadog agent pod in every namespace, and you run custom python checks against your services (e.g. HTTP smokes) from this pod and you do not really need a DaemonSet for it, especially if you need a single pod deployment to save resources and avoid duplication of metrics/events (not only K8s ones, think about your custom ones also).

fiunchinho replied:
Yes, but wouldn't you use the regular daemon set replicas for that? Leaving the Deployment replica just to collect events

ivanilves replied:
* DaemonSet => Global, cluster-wide monitoring
* Deployment => Collect Kubernetes events, optionally provide per-team(per-namespace) endpoint to send custom metrics. So, we may need ports and Service for some cases @fiunchinho

seanknox replied:
Hey @ivanilves, I'm still trying to understand the use case for Deployments. When would an cluster operator opt to provide a per team/namespace datadog metric collection (instead of cluster wide)?

ivanilves replied:
Great question @seanknox !

The first reason is **money**. Each team is a separate organization with it's own Datadog account and API key. They are responsible for their budget, as well as for their custom, application-specific dashboards and alerts.

The second reason is the **flexibility**. Using per-deployment ConfigMap, each team provides it's own settings, it's own "home made" python checks, which are specific to an application run and maintained by a specific team. I mean checks run in "super" namespace make sense for "superApp" run by "superTeam", while they make no sense for "megaTeam" running "megaApp" in "mega" namespace, which is a totally different application, sending totally different metrics, maybe even operating different protocols (one team has XMPP, another - HTTP etc).

*While the global cluster team is a separate unit that just builds clusters and does not care much about application specifics, has its own budget and fine-grained (ha ha, I hope) responsibilities.*

seanknox replied:
> The first reason is money. Each team is a separate organization with it's own Datadog account and API key. They are responsible for their budget, as well as for their custom, application-specific dashboards and alerts.

Thanks for explanation. In what situation does that type of team structure exist? At first blush this seems solving for a very specific team's needs, not a general use case (as is the goal of Helm charts in general).

ivanilves replied:
@seanknox I fully understand your concerns, but I believe it's a pretty general case.

IMO any big (>500 employees), multi-team, multi-unit, budget-aware company could be a good fit for this case. I believe we could call it a general case, because now I see K8s is being adopted even in bigger, conservative/slower companies, with many teams working on projects not having anything in common, but still wanting to enjoy benefits of  cost-saving, unification of the deploy  and management process K8s could provide.

Also please don't forget the **flexibility** part.
Each team could maintain its own DataDog config, adding/changing application-specific checks without touching other teams and global cluster monitoring config. So it's not only about money. :wink:

seanknox replied:
Thanks @ivanilves.

hkaj commented on stable/datadog/templates/deployment.yaml:
```diff
@@ -1,25 +1,37 @@
+        - containerPort: 8126
+          name: traceport
+          protocol: TCP
+        - containerPort: 7777
```
could you remove this port please? It's deprecated

ivanilves replied:
Done @hkaj

ivanilves committed:
Remove legacy port from deployment

ivanilves commented:
Could we somehow test it and, if CI will give OK, merge it? :innocent:  
*I think there is no unaddressed comments anymore here...*

seanknox commented:
@k8s-bot ok to test

fiunchinho commented:
👍 Thank you for the explanation @ivanilves !!

ivanilves commented:
Folks I would like to ask you, were my explanations good enough for you?

Could we merge this PR or it requires more modifications or explanations? 

@lachie83 @gtaylor 
cc: @flah00 @migmartri

vonrabbe reviewed (approved):
Great job @ivanilves

prydonius commented:
lgtm, but want to give @seanknox a chance to check his concerns are addressed.

mazzy89 commented:
In this case DaemonSet and Deployment receive the same `confd`, `checkd` and `autoconf` because the `Configmap`s are the same. Isn't wiser to adopt `Configmap` each one different for Deployment and DaemonSet? This is my use case: 

> I want to specify namespaces list to collect events in the `confd` `kubernetes.yaml` file. In order to do this I create a file and map to a Configmap. The file must be different because the `collectEvents` flag can't be `true` in DaemonSet.

ivanilves commented:
@mazzy89 it depends!

This will introduce more complexity to the single chart install.
In my case I wanted to have only daemonset or deployment 
per chart installation so I did not see a case for doubling configmaps. :wink:

I usually have one daemonset-only installation in default namespace and one deployment-only chart installation per each team-specific namespace.
